src/compiler/checker.ts

```diff
@@ -25251,17 +25251,33 @@ namespace ts {
                 if (isKnownProperty) {
                     // If the check is for a known property (i.e. a property declared in some constituent of
                     // the target type), we filter the target type by presence of absence of the property.
-                    return filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+                    const filtered = filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+                    // without `noUncheckedIndexedAccess` we can't narrow any further
+                    // `{ [x: PropertyKey]: T }[prop]` already returns `T`
+                    // `{ prop?: T }[prop]` always implies `T | undefined`
+                    if (!compilerOptions.noUncheckedIndexedAccess) {
+                        return filtered;
+                    }
+                    return mapType(filtered, t => {
+                        if (getPropertyOfType(type, name)) {
+                            return t;
+                        }
+                        return intersectWithNarrowingRecordType(type, nameType, getApplicableIndexInfoForName(type, name)!.type);
+                    });

```

Comment:
I probably also need to handle optional properties here too - or at least add some tests for them.

Reply:
Oh, I've actually conflated `exactOptionalPropertyTypes` with `noUncheckedIndexedAccess` somehow and I've thought that optional properties would also require some work here. But `noUncheckedIndexedAccess` doesn't affect optional properties anyhow - so the added code looks OK to me now. Also - I've rechecked the current behavior with `exactOptionalPropertyTypes` and without it and it looks OK.